crates/sui-storage/src/event_store/sql.rs

```diff
@@ -357,48 +368,57 @@ const QUERY_BY_OBJECT_ID: &str = "SELECT * FROM events WHERE timestamp >= ? AND
 impl EventStore for SqlEventStore {
     #[instrument(level = "debug", skip_all, err)]
     async fn add_events(&self, events: &[EventEnvelope]) -> Result<u64, SuiError> {
-        // TODO: submit writes in one transaction/batch so it won't just fail in the middle
         let mut cur_seq = self.seq_num.load(Ordering::Acquire);
         let initial_seq = cur_seq;
-        let mut rows_affected: u64 = 0;
-
-        // TODO: benchmark
-        // TODO: use techniques in https://docs.rs/sqlx-core/0.5.13/sqlx_core/query_builder/struct.QueryBuilder.html#method.push_values
-        // to execute all inserts in a single statement?
-        // TODO: See https://kerkour.com/high-performance-rust-with-sqlite
-        for event in events {
-            // Skip events that have a lower sequence number... which must be same or increasing
-            if event.seq_num < cur_seq {
-                debug!(tx_digest =? event.tx_digest, seq_num = event.seq_num, cur_seq, "Skipping event with lower sequence number than current");
+        let mut rows_affected = 0;
+
+        if events.is_empty() {
+            return Ok(0);
+        }
+
+        let mut start_index = 0;
+        let mut end_index = 0;
+        // This while loop only needed to make sure we don't write more than 1000 events at once.
+        // sqlx/SQLite has a limit of binding < 64k elements.
+        while end_index < events.len() {
+            if events[start_index].seq_num < cur_seq {
+                start_index += 1;
+                end_index += 1;
```

Comment:
Do we need to update end_index here? It seems like we reset it outside this if block immediately

Reply:
Yes, because of the `continue`.  There is a case where if end_index is not updated then we could go past `events.len()`